cdap-app-fabric/src/main/java/co/cask/cdap/internal/bootstrap/BootstrapService.java

```diff
@@ -0,0 +1,196 @@
+        } else {
+          bootstrap();
+        }
+      } catch (InterruptedException e) {
+        LOG.info("Bootstrapping could not complete due to interruption. It will be re-run the next time CDAP starts.");
+      }
+    });
+    LOG.info("Started {}", getClass().getSimpleName());
+  }
+
+  @Override
+  protected void shutDown() throws Exception {
+    LOG.info("Stopping {}", getClass().getSimpleName());
+    // Shutdown the executor, which will issue an interrupt to the running thread.
+    // There is only a single daemon thread, so no need to wait for termination
+    executorService.shutdownNow();
+    LOG.info("Stopped {}", getClass().getSimpleName());
+  }
+
+  /**
+   * @return whether the system is bootstrapped or not
+   */
+  public boolean isBootstrapped() {
+    return bootstrapStore.isBootstrapped();
+  }
+
+  /**
+   * Execute all steps in the loaded bootstrap config without skipping any of them.
+   *
+   * @return the result of executing the bootstrap steps.
+   * @throws IllegalStateException if bootstrapping is already in progress
+   * @throws InterruptedException if bootstrapping was interrupted
+   */
+  public BootstrapResult bootstrap() throws InterruptedException {
+    return bootstrap(x -> false);
+  }
+
+  /**
+   * Execute the steps in the loaded bootstrap config.
+   *
+   * @param shouldSkip predicate that determines whether to skip a step
+   * @return the result of executing the bootstrap steps.
+   * @throws IllegalStateException if bootstrapping is already in progress
+   * @throws InterruptedException if bootstrapping was interrupted
+   */
+  public BootstrapResult bootstrap(Predicate<BootstrapStep> shouldSkip) throws InterruptedException {
+    List<BootstrapStepResult> results = new ArrayList<>(config.getSteps().size());
+    if (!bootstrapping.compareAndSet(false, true)) {
+      throw new IllegalStateException("Bootstrap already in progress.");
+    }
+
+    try {
+      for (BootstrapStep step : config.getSteps()) {
+        results.add(executeStep(step, shouldSkip));
+      }
+      if (!config.getSteps().isEmpty()) {
+        bootstrapStore.bootstrapped();
+      }
+      return new BootstrapResult(results);
+    } finally {
+      bootstrapping.set(false);
+    }
+  }
+
+  private BootstrapStepResult executeStep(BootstrapStep step,
+                                          Predicate<BootstrapStep> shouldSkip) throws InterruptedException {
+    try {
+      step.validate();
+    } catch (IllegalArgumentException e) {
+      LOG.warn("Bootstrap step {} failed because it is malformed: {}", step.getLabel(), e.getMessage());
+      return new BootstrapStepResult(step.getLabel(), BootstrapStepResult.Status.FAILED, e.getMessage());
+    }
+
+    if (shouldSkip.test(step)) {
+      return new BootstrapStepResult(step.getLabel(), BootstrapStepResult.Status.SKIPPED, null);
+    }
+
+    BootstrapStepExecutor executor = executors.get(step.getType());
```

Comment:
can we rename the executors to be something else `bootstrapExecutorMap` maybe, it's confusing with `java.util.concurrent.Executors` since you also use ExecutorService in this class..

Reply:
done